### PR TITLE
Bug 1141429 - Improve 'about:home' tabs in the tab drawer

### PR DESCRIPTION
**Pull request by thebnich**
Since about:home shows a view outside of `Browser`, our `screenshot()` function in `Browser` needed some reworking. I created a `ScreenshotHelper` that BVC implements, which we can pass around to classes that need a screenshot without creating a direct dependency to BVC.

### CODE REVIEW

File: Client/Frontend/Browser/TabTrayController.swift [hunk 1]
```diff
@@ -263,15 +258,20 @@ class TabTrayController: UIViewController, UITabBarDelegate, UITableViewDelegate
...
-            let cell = CustomCell(style: UITableViewCellStyle.Default, reuseIdentifier: "id")
-            cell.tab = browser
-            options.moving = cell
-            return cell
+            transitionCell = CustomCell(style: UITableViewCellStyle.Default, reuseIdentifier: "id")
+            options.moving = transitionCell
         }
+
+        if let browser = browser {
+            transitionCell.background.image = screenshotHelper.takeScreenshot(browser, aspectRatio: nil, quality: 1)
```
Comment (wesj) [inline Client/Frontend/Browser/TabTrayController.swift]: If there's no browser passed here, we'll just use whatever image was already in this cell. Is that intended?

File: Client/Frontend/Browser/Browser.swift [hunk 2]
```diff
@@ -45,7 +45,9 @@ class Browser: NSObject, WKScriptMessageHandler {
 
     var title: String? {
         if let title = webView.title {
-        	return title
+            if !title.isEmpty {
+                return title
```
Comment (wesj) [inline Client/Frontend/Browser/Browser.swift]: Unrelated stuff? Also, is an empty title not valid? What do we do with them on Android? I always waffle back and forth on this, but probably title should always return the title and "displayTitle" or something should be used if we want to return the url...

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 3]
```diff
@@ -1160,3 +1163,26 @@ extension BrowserViewController: ReaderModeBarViewDelegate {
...
+
+    init(controller: BrowserViewController) {
+        self.controller = controller
+    }
+
+    func takeScreenshot(tab: Browser, aspectRatio: Float?, quality: Float) -> UIImage? {
+        if let url = tab.url {
+            if url.absoluteString == "about:home" {
+                if let homePanel = controller?.homePanelController {
+                    return homePanel.view.screenshot(aspectRatio: aspectRatio ?? nil, quality: quality)
```
Comment (wesj) [inline Client/Frontend/Browser/BrowserViewController.swift]: Do you need this ?? nil

File: Client/Frontend/Browser/TabTrayController.swift [hunk 4]
```diff
@@ -263,15 +258,20 @@ class TabTrayController: UIViewController, UITabBarDelegate, UITableViewDelegate
...
+        }
+        transitionCell.titleText.text = browser?.title
```
Comment (wesj) [inline Client/Frontend/Browser/TabTrayController.swift]: Ahh, this is why title works they way it does. Still, this should probably be:

browser?.title ?? browser.displayURL?.absoluteString

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 5]
```diff
@@ -1160,3 +1163,26 @@ extension BrowserViewController: ReaderModeBarViewDelegate {
         }
     }
 }
+
+private class BrowserScreenshotHelper: ScreenshotHelper {
+    private weak var controller: BrowserViewController?
+
+    init(controller: BrowserViewController) {
+        self.controller = controller
```
Comment (wesj) [inline Client/Frontend/Browser/BrowserViewController.swift]: I wish we didn't need this controller here, but it'll do for now.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 6]
```diff
@@ -25,6 +25,7 @@ class BrowserViewController: UIViewController {
     private var searchController: SearchViewController?
     private var webViewContainer: UIView!
     private let uriFixup = URIFixup()
+    private var screenshotHelper: ScreenshotHelper!
```
Comment (wesj) [inline Client/Frontend/Browser/BrowserViewController.swift]: Since this isn't a constant, I would probably just make it optional. Unwrapping it here just keeps the compiler from helping us.

File: Client/Frontend/Browser/TabTrayController.swift [hunk 7]
```diff
@@ -161,6 +154,7 @@ class TabTrayController: UIViewController, UITabBarDelegate, UITableViewDelegate
     private let CellIdentifier = "CellIdentifier"
     var tableView: UITableView!
     var profile: Profile!
+    var screenshotHelper: ScreenshotHelper!
```
Comment (wesj) [inline Client/Frontend/Browser/TabTrayController.swift]: This should just be optional too.

File: Client/Frontend/Browser/Browser.swift (hunk 2 above)
Comment (thebnich) [inline Client/Frontend/Browser/Browser.swift]: I wouldn't say it's unrelated (although I did split it into a separate commit) -- it fixes the tab to show "about:home" instead of an empty title in the tab tray. Both desktop and Android show the URL if the title is empty ([test page](https://people.mozilla.org/~bnicholson/test/empty-title.html)). But yeah, `displayTitle` is probably a good idea so we don't mix them up.

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 6 above)
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: I can make it optional, but I'm not sure I get your point. If it's optional and we do forget to set it, the tabs tray will look broken (no thumbnails at all), so we might as well just crash so we can get a stacktrace and fix it.

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 3 above)
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Nope, that doesn't make any sense actually. Thanks.

File: Client/Frontend/Browser/ScreenshotHelper.swift [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+/**
+ * Handles screenshots for a given browser, including pages with non-webview content.
+ */
+protocol ScreenshotHelper {
+    func takeScreenshot(tab: Browser, aspectRatio: Float?, quality: Float) -> UIImage?
```
Comment (st3fan) [inline Client/Frontend/Browser/ScreenshotHelper.swift]: A `nil` aspect ratio is weird and error prone. Why not `aspectRation: Float = 1.0` instead? It is better to choose a good default then to allow nil values imo. Specially for value types like floats.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 9]
```diff
@@ -1160,3 +1163,26 @@ extension BrowserViewController: ReaderModeBarViewDelegate {
...
 }
+
+private class BrowserScreenshotHelper: ScreenshotHelper {
+    private weak var controller: BrowserViewController?
+
+    init(controller: BrowserViewController) {
+        self.controller = controller
+    }
+
+    func takeScreenshot(tab: Browser, aspectRatio: Float?, quality: Float) -> UIImage? {
```
Comment (st3fan) [inline Client/Frontend/Browser/BrowserViewController.swift]: I would turn `aspectRatio` into ` CGFloat` because where you use this you actually convert to `CGFloat` anyway. Everything related to graphics in UIKit uses CGFloat, including point sizes, aspect ratios, positions, etc. (And CGFloat is actually a Float on 32 bit but a Double on 64 bit - but that is an implementation detail)

File: Client/Frontend/Browser/ScreenshotHelper.swift (hunk 8 above)
Comment (thebnich) [inline Client/Frontend/Browser/ScreenshotHelper.swift]: A couple of reasons:
 1. The behavior for `nil` is to use the aspect ratio of the webView. This "default" changes depending on the screen size and layout.
 2. You can't specify defaults in protocols.

I could make the protocol include two different methods, one where `aspectRatio` is missing and another where it's non-optional if you guys think that's cleaner.

File: Client/Frontend/Browser/TabTrayController.swift [hunk 10]
```diff
@@ -250,7 +244,8 @@ class TabTrayController: UIViewController, UITabBarDelegate, UITableViewDelegate
         let tab = tabManager.getTab(indexPath.item)
         let cell = tableView.dequeueReusableCellWithIdentifier(CellIdentifier) as CustomCell
         cell.titleText.text = tab.title
-        cell.background.image = tab.screenshot(size: CGSize(width: tableView.frame.width, height: TabTrayControllerUX.CellHeight))
+        let screenshotAspectRatio = tableView.frame.width / TabTrayControllerUX.CellHeight
+        cell.background.image = screenshotHelper.takeScreenshot(tab, aspectRatio: Float(screenshotAspectRatio), quality: 1)
```
Comment (st3fan) [inline Client/Frontend/Browser/TabTrayController.swift]: I don't think this will work. The problem is that when you open more than about 6 tabs, WebKit will start suspending the content process. And as a result, the screenshot here will be gray. We already had bug reports of that.

Why don't we screenshot, and cache, before the user navigates away from a tab? Then we always have the most recent screenshot of what the user last saw. From a live tab that we know is in good state to screenshot.

File: Utils/ViewExtensions.swift [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+        UIGraphicsBeginImageContextWithOptions(size, false, UIScreen.mainScreen().scale * CGFloat(quality))
+        drawViewHierarchyInRect(CGRect(origin: offset, size: frame.size), afterScreenUpdates: false)
+        var image = UIGraphicsGetImageFromCurrentImageContext()
+        UIGraphicsEndImageContext()
+
+        return image
+    }
+
+    func screenshot(aspectRatio: Float? = nil, offset: CGPoint? = nil, quality: Float = 1) -> UIImage? {
```
Comment (st3fan) [inline Utils/ViewExtensions.swift]: Ah here is the implementation :-) I would turn `aspectRatio` and `quality` into CGFloat types. Then you can get rid of all the casting that you do in both versions of `screenshot()` and things will be equal to UIKit graphics code.

File: Client/Frontend/Browser/TabTrayController.swift (hunk 10 above)
Comment (thebnich) [inline Client/Frontend/Browser/TabTrayController.swift]: Definitely, but that's a different bug! This bug is just about adding about:home support for screenshots.
